### PR TITLE
Avoid impl lookup cycles from evaluating lookup instructions inside an impl decl

### PR DESCRIPTION
**Pull request by danakj**
`LookupImplWitness` instructions inside the impl declaration can't use the impl they are apart of. Previously we had an heuristic in eval which would try to prevent finding the impl for a lookup from inside that impl. But it breaks when the `.Self` is replaced in a generic impl with a symbolic, and then that symbolic is replaced in a specific. The specific's decl block contains that `LookupImplWitness` instruction and it tries to use the impl it came from. This causes the same specific to be formed again, but now it exists, so it's used as-is but it has no decl block yet, and so we crash.

Now we ban an impl while we resolve its specific, both deduction of its arguments and from any other substitution. The prevents instructions from inside the impl (which are evaluated when resolving the specific) from finding their own impl. We do so by adding the ImplId to a stack on the Context, and then skipping such impls when looking for candidates during eval.

This fixes a crash, which was demonstrated by the new test being added. It also makes another todo test pass.

There's a whole lot of other semir churn, which seems to be mostly reordering of constants. There are some fingerprint changes in constants, but it appears they are the same canonical instructions, so they don't represent a behaviour change. For example in `toolchain/check/testdata/for/actual.carbon` the `%N.patt` constant has been given its fingerprint suffix now as `%N.patt.aa5`. But they are both this instruction, so it is just a formatting change:

```
inst6100001A: {kind: SymbolicBindingPattern, arg0: entity_name61000002, type: type(inst61000018)}
  - name: `N`
  - type: type(inst61000018): <pattern for Core.IntLiteral>; {kind: PatternType, arg0: inst(IntLiteralType), type: type(TypeType)} (concrete)
  - value: symbolic_constant61000001
```

### CODE REVIEW

File: toolchain/check/impl_lookup.cpp [hunk 1]
```diff
@@ -721,13 +722,17 @@ static auto CollectCandidateImplsForQuery(
   for (auto [id, impl] : context.impls().enumerate()) {
     CARBON_CHECK(impl.witness_id.has_value());
 
+    // If the impl's interface_id differs from the query, then this impl can
+    // not possibly provide the queried interface.
+    if (impl.interface.interface_id != query_specific_interface.interface_id) {
+      continue;
+    }
```
Comment (cjdb) [inline toolchain/check/impl_lookup.cpp:725]: Why was this moved from its original location?

Comment (danakj) [reply toolchain/check/impl_lookup.cpp:725]: It's cheaper so I moved it up while I was making changes here

File: toolchain/check/context.h [hunk 2]
```diff
@@ -566,6 +570,11 @@ class Context {
   // being checked so that they can be used by later constraints.
   llvm::SmallVector<WhereStackEntry> where_stack_;
 
+  // Impls that can not be used in impl lookup. This is used to prevent cycles
+  // where a lookup inside the impl decl should not be able to find the
+  // containing impl decl.
```
Comment (cjdb) [inline toolchain/check/context.h]:
```suggestion
  // Impls that cannot be used in impl lookup. This prevents cycles where a
  // lookup inside the impl decl should not be able to find the containing impl
  // decl.
```
Nit.

More importantly: I'm not sure what "the containing impl decl" is referring to.

File: toolchain/check/deduce.cpp [hunk 3]
```diff
@@ -676,10 +677,14 @@ auto DeduceImplArguments(Context& context, SemIR::LocId loc_id,
...
-    return SemIR::SpecificId::None;
+  auto specific_id = SemIR::SpecificId::None;
+  if (success) {
+    context.forbidden_impls().push_back(impl_id);
+    specific_id = deduction.MakeSpecific();
+    context.forbidden_impls().pop_back();
   }
-  return deduction.MakeSpecific();
+
+  return specific_id;
```
Comment (cjdb) [inline toolchain/check/deduce.cpp]: I'd prefer to keep the early exit, rather than nest the logic.

File: toolchain/check/testdata/impl/impl_assoc_const.carbon [hunk 4]
```diff
@@ -400,6 +400,26 @@ fn G() {
...
+
+interface I { let X:! type; }
+interface J {}
+
+// When matching this impl during impl lookup, we deduce and replace `T` with a
+// specific argument, and resolve that specific's decl eval block. Doing so
+// re-evaluates the lookup for `I` inside `.X`. If that lookup found this impl
+// (since it's a final impl, eval can find it), we would deduce and produce an
+// infinite cycle. Instructions inside the impl decl should not be able to find
+// the impl.
```
Comment (cjdb) [inline toolchain/check/testdata/impl/impl_assoc_const.carbon]: I'm not sure what this comment block is trying to tell me. Comment blocks like this seem to be you leaving notes for your future self, which I'm okay with. I'm less inclined to make suggestions to these sorts of notes, especially since you eventually remove them. That said, readers should still have an idea of what you're doing here.

Can you provide a short summary sentence at the top, like `Test for X.`?

Comment (danakj) [reply toolchain/check/testdata/impl/impl_assoc_const.carbon]: This is describing the scenario/edge case that the test is testing for. It is describing the process in the toolchain that this test invokes, and what can go wrong that it is verifying does not happen. So this paragraph (as in other tests) is the "Test for X" statement. Do you feel that we should begin writing comments in tests that say something about them being in a test? I don't think I can reduce this paragraph down to a single sentence. Does "Test that we don't cycle in the toolchain." as an intro sentence, when it already says that, change things here? That's not something that we're going to go add to every comment in a test explaining the test case, right? I am having a hard time understanding how to write changes that are going to make sense for review, and still fit into our established patterns and address what I am trying to fix, which I admit is a bit challenging.

File: toolchain/check/context.h (hunk 2 above)
Comment (danakj) [reply toolchain/check/context.h]: The "containing impl decl" is the impl declaration that contains the lookup instruction that this is speaking about.

File: toolchain/check/deduce.cpp (hunk 3 above)
Comment (danakj) [reply toolchain/check/deduce.cpp]: Done

File: toolchain/check/context.h (hunk 2 above)
Comment (danakj) [reply toolchain/check/context.h]: Done

**Review by danakj — COMMENTED** [review]
PTAL

File: toolchain/check/testdata/impl/impl_assoc_const.carbon (hunk 4 above)
Comment (cjdb) [reply toolchain/check/testdata/impl/impl_assoc_const.carbon]: > Does "Test that we don't cycle in the toolchain." as an intro sentence, when it already says that, change things here?

I would find a intro sentence to be helpful. That'll prime me to understand the process under test, and what we expect the outcome to be. I struggled to identify these from your description in isolation.

> That's not something that we're going to go add to every comment in a test explaining the test case, right?

No. I'm requesting a summary for this test because it's a complex unit test. When I see detailed comments, it tells me that the author wants to draw my attention to something subtle. It isn't clear to me that this is testing infinite cycles don't occur: the test name is fairly generic, and infinite cycles are only introduced at the end of the third sentence.

**Review by cjdb — APPROVED** [review]

File: toolchain/check/testdata/impl/impl_assoc_const.carbon (hunk 4 above)
Comment (danakj) [reply toolchain/check/testdata/impl/impl_assoc_const.carbon]: Done

**Comment by danakj** [comment]
I had to make another change here, because in the meantime we landed #7443. It performs a substitution step to thaw out `.Self` references at the end of checking a `where` expression. In this PR's new test we have `I where .X = ()` with a final blanket impl for `I`. This means we get an `ImplWitness` instruction in the access for `.X`, which contains a specific with a single argument for the final impl: `.Self`. When we thaw that `.Self` and rebuild the `ImplWitness` we have a new specific for that final `impl`. Evaluating the `ImplWitness` attempts to resolve the specific for the impl inside it. This is fine - but we need to have the same thing happen here as we were doing in deduce when we formed and resolved the original `impl` specific. The instructions in the `impl`'s specific (which come from inside the `impl` decl) can't do impl lookup and find the `impl` they are inside, or we cycle.

So we need to move the forbidding of the `impl` decl down from deduce into the ResolveSpecificDecl step, and make it conditional on resolving the specific for an `ImplDecl` instruction. Now the new test does not crash again.
